File: frontend/src/types/disease.ts
export type DiseaseStatus = 'Needs Attention' | 'Healthy-looking' | 'Uncertain' | 'Critical';

export interface DiseasePrediction {
  diseaseName: string;
  confidence: number; // percentage, e.g. 91
  description?: string;
}

export interface DiseaseResult {
  id: string;
  cropName: string;
  selectedCrop?: string;
  detectedCrop?: string;
  cropConfidence?: number;
  cropMatch?: boolean;
  predictionAllowed?: boolean;
  isCropMismatch?: boolean;
  errorCode?: string;
  primaryCondition: string;
  confidence: number;
  status: DiseaseStatus;
  isLowConfidence: boolean;
  isHealthy: boolean;
  visualFindings: string;
  nextSteps: string[];
  topPredictions: DiseasePrediction[];
  scannedAt: string;
  imageUrl: string;
  farmId?: string;
  farmName?: string;
}

export interface ImageQualityCheck {
  isValid: boolean;
  lightingQuality: 'Good' | 'Fair' | 'Poor';
  blurLevel: 'Low' | 'Moderate' | 'High';
  isLeafCentered: boolean;
  notes: string;
}

export interface BackendPredictResponse {
  success: boolean;
  status: 'success' | 'low_confidence' | 'unsupported_image' | 'crop_mismatch' | 'error';
  selectedCrop?: string | null;
  detectedCrop?: string | null;
  cropConfidence?: number | null;
  cropMatch?: boolean | null;
  predictionAllowed?: boolean | null;
  errorCode?: string | null;
  crop?: string | null;
  disease?: string | null;
  class_name?: string | null;
  confidence: number;
  message?: string | null;
}

File: backend/app/schemas/disease.py
from typing import Optional, List, Literal
from pydantic import BaseModel, Field, ConfigDict

DiseaseStatusType = Literal["Needs Attention", "Healthy-looking", "Uncertain", "Critical"]


class DiseasePredictionItem(BaseModel):
    diseaseName: str = Field(..., serialization_alias="diseaseName")
    confidence: float  # e.g. 91.0 (%)
    description: Optional[str] = None
    model_config = ConfigDict(populate_by_name=True)


class DiseaseResultResponse(BaseModel):
    id: str
    cropName: str = Field(..., serialization_alias="cropName")
    primaryCondition: str = Field(..., serialization_alias="primaryCondition")
    confidence: float
    status: DiseaseStatusType
    isLowConfidence: bool = Field(..., serialization_alias="isLowConfidence")
    isHealthy: bool = Field(..., serialization_alias="isHealthy")
    visualFindings: str = Field(..., serialization_alias="visualFindings")
    nextSteps: List[str] = Field(default_factory=list, serialization_alias="nextSteps")
    topPredictions: List[DiseasePredictionItem] = Field(default_factory=list, serialization_alias="topPredictions")
    scannedAt: str = Field(..., serialization_alias="scannedAt")
    imageUrl: str = Field(..., serialization_alias="imageUrl")
    farmId: Optional[str] = Field(None, serialization_alias="farmId")
    farmName: Optional[str] = Field(None, serialization_alias="farmName")
    model_version: Optional[str] = "mobilenetv2-v1.0.0"

    model_config = ConfigDict(populate_by_name=True)


class ImageQualityCheck(BaseModel):
    isValid: bool = Field(..., serialization_alias="isValid")
    lightingQuality: Literal["Good", "Fair", "Poor"] = Field(..., serialization_alias="lightingQuality")
    blurLevel: Literal["Low", "Moderate", "High"] = Field(..., serialization_alias="blurLevel")
    isLeafCentered: bool = Field(..., serialization_alias="isLeafCentered")
    notes: str

    model_config = ConfigDict(populate_by_name=True)


class PredictResponse(BaseModel):
    success: bool = True
    status: Literal["success", "low_confidence", "unsupported_image", "error"] = "success"
    crop: Optional[str] = None
    disease: Optional[str] = None
    class_name: Optional[str] = None
    confidence: float
    message: Optional[str] = None

    model_config = ConfigDict(populate_by_name=True)

